Implement patterns

rerobika opened the pull request:
Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

yichoi commented:
@rerobika please rebase this PR to the master

yichoi commented:
@rerobika rebase please

clover2123 commented:
@rerobika It may be too late, but would you fix `src/parser/esprima_cpp/esprima.cpp` according to the [esprima](https://github.com/jquery/esprima.git) (#commit_id cfd8eba831517cc89d895eb0e992ec78210a8a0d) as much as possible?
Since our parser is implemented based on the `esprima`, matching to it is highly recommended. It is also good for updating/maintenance.

rerobika commented:
@clover2123 I've rebased the PR to the latest master and updated the parsing state of the object/array patterns to stay synchronized with the commit that you mentioned.

clover2123 commented on src/parser/ast/ArrayPatternNode.h:
```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
```
It's a minor fix, 2016 -> 2019

clover2123 commented on src/parser/ast/ObjectPatternNode.h:
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
```
Same as above.

clover2123 commented on src/parser/ast/RegisterReferenceNode.h:
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
```
The last one. Other looks nice to me!

rerobika replied:
Thanks, I fixed it.
